libsolidity/lsp/LanguageServer.cpp

```diff
@@ -0,0 +1,354 @@
+}
+
+bool LanguageServer::clientPathSourceKnown(string const& _path) const
+{
+	return m_fileReader.sourceCodes().count(clientPathToSourceUnitName(_path));
+}
+
+void LanguageServer::changeConfiguration(Json::Value const& _settings)
+{
+	m_settingsObject = _settings;
+}
+
+void LanguageServer::compile()
+{
+	// TODO: optimize! do not recompile if nothing has changed (file(s) not flagged dirty).
+
+	m_compilerStack.reset(false);
+	m_compilerStack.setSources(m_fileReader.sourceCodes());
+	m_compilerStack.compile(CompilerStack::State::AnalysisPerformed);
+}
+
+void LanguageServer::compileAndUpdateDiagnostics()
+{
+	compile();
+
+	map<string, Json::Value> diagnosticsBySourceUnit;
+	for (string const& sourceUnitName: m_fileReader.sourceCodes() | ranges::views::keys)
+		diagnosticsBySourceUnit[sourceUnitName] = Json::arrayValue;
+
+	for (shared_ptr<Error const> const& error: m_compilerStack.errors())
+	{
+		SourceLocation const* location = error->sourceLocation();
+		if (!location || !location->sourceName)
+			// LSP only has diagnostics applied to individual files.
+			continue;
+
+		Json::Value jsonDiag;
+		jsonDiag["source"] = "solc";
+		jsonDiag["severity"] = toDiagnosticSeverity(error->type());
+		jsonDiag["code"] = Json::UInt64{error->errorId().error};
+		string message = error->typeName() + ":";
+		if (string const* comment = error->comment())
+			message += " " + *comment;
+		jsonDiag["message"] = move(message);
+		jsonDiag["range"] = toRange(*location);
+
+		if (auto const* secondary = error->secondarySourceLocation())
+			for (auto&& [secondaryMessage, secondaryLocation]: secondary->infos)
+			{
+				Json::Value jsonRelated;
+				jsonRelated["message"] = secondaryMessage;
+				jsonRelated["location"] = toJson(secondaryLocation);
+				jsonDiag["relatedInformation"].append(jsonRelated);
+			}
+
+		diagnosticsBySourceUnit[*location->sourceName].append(jsonDiag);
+	}
+
+	for (string const& sourceUnitName: m_fileReader.sourceCodes() | ranges::views::keys)
+	{
+		Json::Value params;
+		params["uri"] = sourceUnitNameToClientPath(sourceUnitName);
+		params["diagnostics"] = move(diagnosticsBySourceUnit.at(sourceUnitName));
+		m_client.notify("textDocument/publishDiagnostics", move(params));
+	}
+}
+
+bool LanguageServer::run()
+{
+	while (!m_exitRequested && !m_client.closed())
+	{
+		optional<Json::Value> const jsonMessage = m_client.receive();
+		if (!jsonMessage)
+		{
+			m_client.error({}, ErrorCode::ParseError, "Error parsing JSONRPC request.");
+			continue;
+		}
+
+		try
+		{
+			string const methodName = (*jsonMessage)["method"].asString();
+			MessageID const id = (*jsonMessage)["id"];
+
+			if (auto handler = valueOrDefault(m_handlers, methodName))
+				handler(id, (*jsonMessage)["params"]);
+			else
+				m_client.error(id, ErrorCode::MethodNotFound, "Unknown method " + methodName);
+		}
+		catch (exception const& e)
+		{
+			m_client.error({}, ErrorCode::InternalError, "Unhandled exception: "s + e.what());
```

Comment:
These are all cases where some part of the protocol is not correctly implemented. I think for a first release it is fine to keep them short. If this will result in bug reports being complicated to dig into, we can extend it later on.

Reply:
But getting diagnostic info from boost is a one-liner and would immediately give us much more info. I don't see why we should not make it more informative from the start if it's easy. Bad error handling that swallows too much info is very annoying.